### DISCUSSION TITLE
Zapier / Stripe - Fatture In Cloud. Problema campo data.entity.contry + numerazione fatture

### DISCUSSION THREAD

**Discussion by Dreamerscommunity** [fattureincloud/api#466] [Q&A] [answered]
Ciao @valmoz, e a tutta la Community. spero tutto bene.

Ho quasi terminato il setup della fatturazione automatica su Fatture in cloud tramite Zapier, ma purtroppo riscontro un errore che non riesco a risolvere
![image](https://github.com/user-attachments/assets/9670c3d8-120a-45d9-9ba1-d74fd05aa435)

Mi dice "data.entity.country" field is not valid. Attualmente lo passiamo come "IT" (Come da tracciato del pagamento zapier).

Il supporto Zapier non riesce aiutarmi. Hai idea di come potrei risolvere?

Un'altra domanda: sapresti dirmi come potrei far in modo che Fatture in Cloud emetta una fattura con la numerazione che vogliamo noi? Grazie

Marco

**Comment by valmoz** [comment, marked as answer]
Buongiorno.
Il campo **country** accetta il nome del campo per esteso, come ad esempio "Italia". Può trovare l'elenco delle Countries effettuando la chiamata [List Countries](https://developers.fattureincloud.it/api-reference#get-/info/countries).

Per poter utilizzare il valore "IT" che riceve dal tracciato del pagamento Zapier deve usare invece il campo **country_iso**.
Può trovare l'elenco dei valori accettati con la chiamata [List Detailed Countries](https://developers.fattureincloud.it/api-reference#get-/info/detailed_countries).

Come spiegato [qui](https://developers.fattureincloud.it/docs/guides/invoice-creation#two-step-two-the-invoice-data), il campo **number** permette di definire il numero della fattura, escludendo la numerazione automatica effettuata da Fatture in Cloud.

Le suggerisco di seguire la guida per la creazione di fatture, che include anche degli esempi Zapier: [Invoice Creation](https://developers.fattureincloud.it/docs/guides/invoice-creation)

Grazie
Mauro

**Reply by Dreamerscommunity** [reply]
Buonasera @valmoz ho risolto per il campo Country, grazie mille!

Invece per il numero progressivo, io vorrei che fosse M-(numero progressivo), come potrei impostarlo?

Grazie per la pazienza, scusami ma sono molto poco tecnico su questa parte :)

Marco

**Reply by valmoz** [reply]
@Dreamerscommunity non è possibile farlo.
Può però utilizzare il campo **numeration** per definire il sezionale, come indicato dalla guida di creazione fattura.

Grazie
Mauro

**Comment by Dreamerscommunity** [comment]
Buon pomeriggio @valmoz 
Quindi sostanzialmente mi dici che non posso selezionare il numero liberamente partendo da un progressivo che deciderò io, ma che posso almeno inserire il sezionale, cioè **M-** , corretto?
Ti potrei richiedere un esempio di come potrei settarlo?  Grazie infinite

**Reply by valmoz** [reply]
Buongiorno @Dreamerscommunity 
come spiegato qui sopra, può usare il campo **number** per definire il numero di fattura che le serve e **numeration** per definire il sezionale. I due campi possono essere usati in combinazione.

Ad esempio:
```json
{
  "data": {
    "type": "invoice",
    "entity": {
      ...
    },
    "number": 1234,
    "numeration": "M",
   ...
  }
}
```

Nella guida [Invoice Creation](https://developers.fattureincloud.it/docs/guides/invoice-creation#two-step-two-the-invoice-data) trova già degli esempi anche per Zapier, quindi le suggerisco di prenderne visione.

Per effettuare dei test può anche provare ad utilizzare la nostra [API Reference](https://developers.fattureincloud.it/api-reference) prima di provare con Zapier, in modo tale da aver ben chiaro cosa è necessario implementare per ottenere quanto richiesto dal suo caso d'uso. Qui una guida che spiega il funzionamento dell'[API Reference](https://developers.fattureincloud.it/docs/api-reference).

Le ricordo che finché non verrà utilizzato il metodo Send E-Invoice sarà sempre possibile cancellare o modificare i documenti, in modo tale da poter effettuare tutti i test che desidera. Una volta inviato al SDI invece il documento non può più essere modificato. Più informazioni qui: [E-Invoice Management](https://developers.fattureincloud.it/docs/guides/e-invoice-management)

Grazie
Mauro

**Comment by Dreamerscommunity** [comment]
Buongiorno @valmoz 

Così può andare? Il campo numeration è compilato con il formato che vorrei. (M-)
![image](https://github.com/user-attachments/assets/4529fe6b-5992-4f0f-9cc4-f35e92ec5672)


Il campo number invece è lasciato volutamente vuoto in modo che prenda il prossimo disponibile

Grazie

**Reply by valmoz** [reply]
Buongiorno,
le consiglio di effettuare una prova e verificare il risultato.
Le ricordo che non sarà comunque possibile ottenere il formato "M-1234", per questo le consiglio di effettuare una prova per verificate il risultato ottenuto.

Per la prova dovrà disabilitare temporaneamente la chiamata Send E-Invoice, o in alternativa impostare a "true" il campo [dry_run](https://developers.fattureincloud.it/docs/guides/e-invoice-management#pirate_flag-the-dry-run-flag).

Mauro

**Comment by Dreamerscommunity** [comment]
@valmoz grazie, il dry run è impostato su TRUE, quindi posso fare i test del caso.

Un'altra domanda: lo step "Get invoice XML" in realtà è superfluo, e posso eliminarlo, passando direttamente a "Send E-Invoice"; corretto?

**Reply by valmoz** [reply]
corretto

**Comment by Dreamerscommunity** [comment]
Ciao @valmoz 
Anche la numerazione ora funziona come vorrei. Mi restano solo 2 problematiche da risolvere e poi direi che ci siamo.

Sapresti dirmi nel tracciato di Stripe, come posso prendere i valori dell'imponibile, dell'IVA, e del totale?


Oltre a dove prendere avrei bisogno di sapere poi dove scriverli su Fatture in cloud.

L'obiettivo è avere come output una fattura così generata su Fatture in Cloud (questa l'ho fatta io manualmente)

<img width="560" alt="image" src="https://github.com/user-attachments/assets/ba134b60-0ea9-4409-b272-56dfa8873817">


Grazie davvero,

Marco

**Reply by valmoz** [reply]
Buongiorno @Dreamerscommunity 
Per quanto riguarda i campi su Fatture in Cloud, trova tutto scritto sulla guida [Invoice Creation](https://developers.fattureincloud.it/docs/guides/invoice-creation#two-step-two-the-invoice-data), che è stata creata per guidarla passo passo nella creazione di una fattura tramite API.

Avendo lei creato una fattura d'esempio su Fatture in Cloud, in alternativa può anche provare ad effettuare una chiamata [Get Issued Document](https://developers.fattureincloud.it/api-reference#get-/c/-company_id-/issued_documents/-document_id-) o una [List Issued Documents](https://developers.fattureincloud.it/api-reference#get-/c/-company_id-/issued_documents) dalla nostra [API Reference](https://developers.fattureincloud.it/docs/api-reference) e vedere il risultato restituito; il risultato fornito da questi metodi corrisponde quasi esattamente al body che dovrà essere passato alla richiesta per poter ottenere la fattura così come la vuole lei (al netto degli identificativi assegnati in fase di creazione, che non possono essere conosciuti a priori).

Per quanto riguarda il tracciato Stripe, purtroppo non la posso aiutare 😄 Credo che le convenga contattare il supporto di Stripe.

Grazie
Mauro

**Comment by Dreamerscommunity** [comment]
Buongiorno @valmoz 

Ho sentito Stripe e sto aspettando che mi dicano dove trovare i campi nel tracciato; al momento ho recuperato solo quello corrispondente a quello che in Fatture in Cloud è "issued.document item gross price", cioè il totale lordo.

Ti faccio però una domanda: io ho in vendita un solo prodotto, quindi diciamo le fatture saranno sempre identiche: 15€ di lordo = 12,70€ netto + 2,30€ di IVA 22%. Secondo te, posso "Forzare" i campi, scrivendoli a mano, anzichè prendendoli dal tracciato Stripe? 

Grazie

Marco

**Reply by Dreamerscommunity** [reply]
Ciao @valmoz sto intanto provando in qualche modo a fare dei test, ma non ne vengo fuori :(

L'errore principale che riscontro è "Failed to create an issued documents in Fatture in Cloud
Il totale dei pagamenti non corrisponde al totale da pagare."

Ti copio la lista dei campi che ho compilato. Potresti dirmi cosa non torna?

![image](https://github.com/user-attachments/assets/6384b02f-12d4-46f2-af5d-719d5997fecb)


Inoltre volevo chiederti: sul campo data.payments_list.due_date io ho forzato scrivendo a mano la data di questa fattura, giusto per vedere se funzionava, perchè dal tracciato stripe, la data transazione esce con l'ora e non viene accettata da fatture in cloud. C'è modo di scrivere qualcosa sul campo in modo che di default prenda la data in cui viene triggerato il flusso?

Grazie di nuovo

Marco

**Reply by valmoz** [reply]
Buongiorno @Dreamerscommunity 
per quanto riguarda l'errore che sta ricevendo, probabilmente si tratta di una questione di approssimazioni.
La guida in cui spiega come risolvere è questa: [Invoice Totals Calculation](https://developers.fattureincloud.it/docs/guides/invoice-totals), ma poiché lei avrà solo un prodotto sempre uguale le suggerisco nuovamente di provare a recuperare i dettagli della sua fattura d'esempio, come suggerito in precedenza, e di utilizzare esattamente gli stessi valori usandoli come costanti in Zapier.

Per quanto riguarda la due_date, purtroppo è obbligatoria. Le chiederei di verificare con il supporto di Zapier se è possibile ottenere una data nel formato YYYY-MM-DD negli step precedenti.

Mauro

**Reply by valmoz** [reply]
Ad esempio, penso che potrebbe provare a combinare le seguenti funzionalità di Zapier:
- [Insert the time your Zap runs into a field](https://help.zapier.com/hc/en-us/articles/8496275717261-Insert-the-time-your-Zap-runs-into-a-field#h_01HJ1D5RZTSWA5DD5F4JHSJYYG)
- [Modify date and time formats in Zaps](https://help.zapier.com/hc/en-us/articles/8496257974029-Modify-date-and-time-formats-in-Zaps#h_01HF6R8YMZGNMZD1K4V55RZYGN)

Altri articoli che potrebbero darle una mano:
- [How to change date and time formats in your text automatically](https://zapier.com/blog/format-datetimes/)
- [Formatter Features: Date Formatting](https://community.zapier.com/featured-articles-65/formatter-features-date-formatting-10809)

**Reply by Dreamerscommunity** [reply]
Buon pomeriggio @valmoz 

Per la data, mi informo con il supporto di zapier e spero di trovare una soluzione.

Mentre per i totali che non tornano, dici che sia perchè per fare i 15€ totali, sulla fattura che abbiamo emesso a mano, abbiamo dovuto fare uno sconto di 0,1 cent, come nell'immagine?

![image](https://github.com/user-attachments/assets/f9d99fcf-23dd-4cf0-bee7-82139dce00dd)


Abbiamo notato emettendola a mano, che effettivamente FIC aveva necessità di questo sconto per arrivare alla cifra totale lorda di 15€. Se è questo il motivo, sapresti dirmi come posso applicare lo sconto di default?

Non sono in grado di richiamare il documento con "Get issued document" dalla API reference purtroppo.

Grazie

Marco

**Reply by valmoz** [reply]
Buongiorno @Dreamerscommunity 
Le consiglio piuttosto di utilizzare i gross_price invece dei net_price, andando a specificare il solo lordo invece del netto. In questo modo dovrebbe ottenere 15€ tondi senza necessità di aggiungere sconti.

Se necessario, possiamo recuperare noi la fattura d'esempio tramite il Supporto Clienti. Può contattare il nostro Supporto Clienti tramite info@fattureincloud.it e allegando il link a questa discussione. 

Grazie
Mauro

**Reply by Dreamerscommunity** [reply]
Buongiorno @valmoz 
Usando il gross price ora tutto funziona come vorrei. L'unica cosa che mancherebbe, è il codice iva giusto (data.items_list.vad.it). Ho utilizzato "1" ma mi riporta IVA al 21% anzichè al 22%. Puoi aiutarmi a recuperare quello corretto per il 22%?

Grazie infinite

Marco

**Reply by valmoz** [reply]
@Dreamerscommunity 
come spiegato [qui](https://developers.fattureincloud.it/docs/guides/invoice-creation#five-step-five-items-list) può usare il metodo [List Vat Types](https://developers.fattureincloud.it/api-reference/#get-/c/-company_id-/info/vat_types) per recuperare l'elenco dei tipi a disposizione.
In ogni caso, a memoria direi che l'ID corretto è "0".

**Comment by Dreamerscommunity** [comment]
@valmoz 
Ottima memoria, ha funzionato :) Grazie!

L'ultima cosa che manca e poi davvero ci siamo è lo status del pagamento. Vorrei che risultasse già pagata (Sono tutti pagamenti con Carta di Credito) ma ricevo questo errore

![image](https://github.com/user-attachments/assets/5f014b04-752d-40d8-b6ae-3c53d40233bb)

![image](https://github.com/user-attachments/assets/dee76a8c-8a01-4aef-8832-16818c8a41f0)


Quale campo manca?

Grazie

Marco

**Reply by valmoz** [reply]
@Dreamerscommunity 
Come può leggere [qui](https://developers.fattureincloud.it/docs/guides/invoice-creation#six-step-six-payments-list)
![image](https://github.com/user-attachments/assets/4258e64d-9da8-4b21-a76d-0c8b263cb34f)
Può creare un Payment Account (magari l'ha già creato dal sito per creare la fattura di test) e può utilizzarlo all'interno del payment.

Per trovare gli ID degli account esistenti, deve usare una [List Payments Accounts](https://developers.fattureincloud.it/api-reference#get-/c/-company_id-/info/payment_accounts).

**Reply by Dreamerscommunity** [reply]
@Dreamerscommunity grazie, per recuperare gli ID degli account esistenti tramite "List Payments Accounts" devo inserire un company ID

![image](https://github.com/user-attachments/assets/80e1e24a-9e71-44e3-b5a5-21d99b414311)

Dove trovo questa informazione?

**Reply by valmoz** [reply]
@Dreamerscommunity
controlli qui: [Where can I find my Company ID?](https://developers.fattureincloud.it/docs/basics/company-scoped-methods#microscope-where-can-i-find-my-company_id)

credo però che possa vederlo anche da Zapier, in quanto al momento della selezione della company da usare per un'azione specifica dovrebbe mostrare anche gli ID selezionati.

In alternativa può anche scrivere al Supporto Clienti (tramite info@fattureincloud.it) allegando il link a questa discussione.

Mauro

**Reply by Dreamerscommunity** [reply]
@valmoz ho scritto al support. Ho trovato l'ID su Zapier (Credo) ma il sistema "List Payment Accounts", una volta inserito l'ID, mi risponde "Unauthorized"

Grazie

**Reply by valmoz** [reply]
@Dreamerscommunity significa che non ha inserito il token per l'[autenticazione](https://developers.fattureincloud.it/docs/basics/authentication).

Le consiglio di attendere risposta dal supporto, e poi eventualmente di procedere a create un token come indicato. Credo che dal Supporto saranno in grado di aiutarla.

Mauro

**Reply by valmoz** [reply]
Buongiorno @Dreamerscommunity è riuscito a risolvere con il supporto?

Grazie
Mauro

**Reply by Dreamerscommunity** [reply]
Buongiorno @valmoz confermo, tutto risolto. Si può chiudere il thread :)

Grazie di tutto

Marco
